File: app/buildz/tmp/kapt3/stubs/debug/com/bits/bee/bpmc/data/repository/UserRepositoryImpl.java
package com.bits.bee.bpmc.data.repository;

import java.lang.System;

/**
 * Created by aldi on 29/03/22.
 */
@kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u00006\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010 \n\u0002\b\u0003\n\u0002\u0010\b\n\u0000\u0018\u00002\u00020\u0001B\u0017\b\u0007\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u0012\u0006\u0010\u0004\u001a\u00020\u0005\u00a2\u0006\u0002\u0010\u0006J\u0010\u0010\u0007\u001a\n\u0012\u0006\u0012\u0004\u0018\u00010\t0\bH\u0016J\u0014\u0010\n\u001a\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\t0\u000b0\bH\u0016J#\u0010\f\u001a\u0014\u0012\u0010\u0012\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\t0\r0\u000b0\bH\u0096@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\u000eJ\u001c\u0010\u000f\u001a\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\t0\u000b0\b2\u0006\u0010\u0010\u001a\u00020\u0011H\u0016R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0002\u001a\u00020\u0003X\u0082\u0004\u00a2\u0006\u0002\n\u0000\u0082\u0002\u0004\n\u0002\b\u0019\u00a8\u0006\u0012"}, d2 = {"Lcom/bits/bee/bpmc/data/repository/UserRepositoryImpl;", "Lcom/bits/bee/bpmc/domain/repository/UserRepository;", "userDao", "Lcom/bits/bee/bpmc/data/data_source/local/dao/UserDao;", "defaultDispatcher", "Lkotlinx/coroutines/CoroutineDispatcher;", "(Lcom/bits/bee/bpmc/data/data_source/local/dao/UserDao;Lkotlinx/coroutines/CoroutineDispatcher;)V", "getActiveUser", "Lkotlinx/coroutines/flow/Flow;", "Lcom/bits/bee/bpmc/domain/model/User;", "getDefaultUser", "Lcom/bits/bee/bpmc/utils/Resource;", "getUser", "", "(Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "getUserById", "id", "", "app_debug"})
public final class UserRepositoryImpl implements com.bits.bee.bpmc.domain.repository.UserRepository {
    private final com.bits.bee.bpmc.data.data_source.local.dao.UserDao userDao = null;
    private final kotlinx.coroutines.CoroutineDispatcher defaultDispatcher = null;
    
    @javax.inject.Inject()
    public UserRepositoryImpl(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.data.data_source.local.dao.UserDao userDao, @org.jetbrains.annotations.NotNull()
    kotlinx.coroutines.CoroutineDispatcher defaultDispatcher) {
        super();
    }
    
    @org.jetbrains.annotations.Nullable()
    @java.lang.Override()
    public java.lang.Object getUser(@org.jetbrains.annotations.NotNull()
    kotlin.coroutines.Continuation<? super kotlinx.coroutines.flow.Flow<? extends com.bits.bee.bpmc.utils.Resource<? extends java.util.List<com.bits.bee.bpmc.domain.model.User>>>> continuation) {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.domain.model.User> getActiveUser() {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<com.bits.bee.bpmc.domain.model.User>> getDefaultUser() {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<com.bits.bee.bpmc.domain.model.User>> getUserById(int id) {
        return null;
    }
}

File: app/src/main/java/com/bits/bee/bpmc/data/repository/UserRepositoryImpl.kt
package com.bits.bee.bpmc.data.repository

import com.bits.bee.bpmc.data.data_source.local.dao.UserDao
import com.bits.bee.bpmc.data.data_source.remote.ApiUtils
import com.bits.bee.bpmc.data.data_source.remote.response.BranchResponse
import com.bits.bee.bpmc.data.data_source.remote.response.BranchUserResponse
import com.bits.bee.bpmc.data.data_source.remote.response.InitialResponse
import com.bits.bee.bpmc.domain.mapper.UserDataMapper
import com.bits.bee.bpmc.domain.model.Branch
import com.bits.bee.bpmc.domain.model.User
import com.bits.bee.bpmc.domain.repository.UserRepository
import com.bits.bee.bpmc.utils.ApiResponse
import com.bits.bee.bpmc.utils.NetworkDatabaseBoundResource
import com.bits.bee.bpmc.utils.Resource
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.withContext
import javax.inject.Inject


/**
 * Created by aldi on 29/03/22.
 */
class UserRepositoryImpl @Inject constructor(
    private val userDao: UserDao,
    private val defaultDispatcher: CoroutineDispatcher,
    private val apiUtils: ApiUtils
) : UserRepository {

    override suspend fun getUser(): Flow<Resource<List<User>>> = flow<Resource<List<User>>> {
//        userDao.deleteAll()

    }.flowOn(defaultDispatcher)

    override fun getActiveUser(): Flow<User?> {
        return flow {
            val data = userDao.getDefaultUser()
            emit(data?.let { UserDataMapper.fromDbToDomain(it) })
        }.flowOn(defaultDispatcher)
    }

    override fun getUserById(id: Int): Flow<User> {
        return flow {
            emit(UserDataMapper.fromDbToDomain(userDao.getUserById(id)))
        }.flowOn(defaultDispatcher)
    }

    override fun getBranchUser(): Flow<Resource<Any>> {
        return object : NetworkDatabaseBoundResource<Any, InitialResponse>(){
            override suspend fun loadFormDB(): Any? = Any()

            override fun shouldFetch(data: Any?): Boolean = true

            override suspend fun createCall(): Flow<ApiResponse<InitialResponse>> {
                return apiUtils.getInitialApiService().getInitial()
            }

            override suspend fun saveCallResult(data: InitialResponse) {
                val valdata = data.data
                userDao.insertBulk(valdata.usr.map { UserDataMapper.fromNetworkToDb(it)})
            }

        }.getAsFlow()
    }

    override fun getUserPin(user: String): Flow<List<User>> {
        return flow {
            val data = userDao.getUserPin(user).map { UserDataMapper.fromDbToDomain(it) }
            emit(data)
        }.flowOn(defaultDispatcher)
    }

    override fun getUserBySecretSauce(
        username: String,
        userapi: String
    ): Flow<Resource<List<User>>> {
        return flow {
            val data = userDao.getUserBySecretSauce(username, userapi).map { UserDataMapper.fromDbToDomain(it) }
            emit(Resource.success(data))
        }.flowOn(defaultDispatcher)
    }

    override suspend fun updateUser(user: User) {
        withContext(defaultDispatcher){
            userDao.update(UserDataMapper.fromDomainToDb(user))
        }
    }

    override suspend fun resetUsedUser() {
        withContext(defaultDispatcher){
            userDao.resetUsed()
        }
    }
}